pkg/queue/timeout.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+)
+
+var defaultTimeoutBody = "<html><head><title>Timeout</title></head><body><h1>Timeout</h1></body></html>"
+
+// TimeToFirstByteTimeoutHandler returns a Handler that runs h with the given time limit
+// in which the first byte of the response must be written.
+//
+// The new Handler calls h.ServeHTTP to handle each request, but if a
+// call runs for longer than its time limit, the handler responds with
+// a 503 Service Unavailable error and the given message in its body.
+// (If msg is empty, a suitable default message will be sent.)
+// After such a timeout, writes by h to its ResponseWriter will return
+// ErrHandlerTimeout.
+//
+// The implementation is largely inspired by http.TimeoutHandler.
+func TimeToFirstByteTimeoutHandler(h http.Handler, dt time.Duration, msg string) http.Handler {
+	return &timeoutHandler{
+		handler: h,
+		body:    msg,
+		dt:      dt,
+	}
+}
+
+type timeoutHandler struct {
+	handler http.Handler
+	body    string
+	dt      time.Duration
+}
+
+func (h *timeoutHandler) errorBody() string {
+	if h.body != "" {
+		return h.body
+	}
+	return defaultTimeoutBody
+}
+
+func (h *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx, cancelCtx := context.WithCancel(r.Context())
+	defer cancelCtx()
+
+	done := make(chan struct{})
+	tw := &timeoutWriter{w: w}
+	panicChan := make(chan interface{}, 1)
+	go func() {
+		defer func() {
+			if p := recover(); p != nil {
+				panicChan <- p
+			}
+		}()
+		h.handler.ServeHTTP(tw, r.WithContext(ctx))
+		close(done)
+	}()
+
+	timeout := time.After(h.dt)
+	for {
+		select {
+		case p := <-panicChan:
+			panic(p)
```

Comment:
Can you document this a bunch more?  Why is it using `panic`s in this way?

Reply:
I can only interpret here, because this part of it is taken from the original timeout handler. I think though, we need to propagate panics properly because the server handles panicking ServeHTTP in a specific way. I'll do a bit of a readup and add some infomation to the overall comment.